test/cql-pytest/test_materialized_view.py

```diff
@@ -1354,3 +1354,56 @@ def test_create_materialized_view_slash_name(cql, test_keyspace, table1):
         # We shouldn't reach here, but if we did, let the test fail cleanly
         cql.execute(f'DROP MATERIALIZED VIEW IF EXISTS {test_keyspace}."/xyz/"')
+
+@pytest.fixture(scope="module")
+def mv1(cql, test_keyspace):
+    with new_test_table(cql, test_keyspace, 'p int, c int, x int, primary key (p)') as table:
+        with new_materialized_view(cql, table, '*', 'c, p', 'p is not null and c is not null') as mv:
+            yield mv
+
+# A materialized view cannot be written to directly - it can only be written
+# through its base table. This test verifies that all CQL operations that can
+# write to a table - INSERT, UPDATE, BATCH, DELETE, and TRUNCATE - are
+# rejected on a view table.
+def test_view_forbids_write(cql, mv1):
+    with pytest.raises(InvalidRequest, match='Cannot directly modify a materialized view'):
+        cql.execute(f'INSERT INTO {mv1} (c,p) VALUES (1,2)')
+    with pytest.raises(InvalidRequest, match=f'Cannot directly modify a materialized view'):
+        cql.execute(f'UPDATE {mv1} SET x=1 WHERE c=1 AND p=2')
+    with pytest.raises(InvalidRequest, match='Cannot directly modify a materialized view'):
+        cql.execute(f'BEGIN BATCH UPDATE {mv1} SET x=1 WHERE c=1 AND p=2; APPLY BATCH')
+    with pytest.raises(InvalidRequest, match='Cannot directly modify a materialized view'):
+        cql.execute(f'DELETE FROM {mv1} WHERE c=1 AND p=2')
+    with pytest.raises(InvalidRequest, match='Cannot TRUNCATE materialized view directly'):
+        cql.execute(f'TRUNCATE {mv1}')
+
+# A materialized view should not be operated on with operations that have
+# "TABLE" in their name - DROP TABLE, ALTER TABLE, and DESC TABLE. There
+# are identical operations with "MATERIALIZED VIEW" in their name, which
+# should be used instead.
+@pytest.mark.xfail(reason="issue #21026")
+def test_view_forbids_table_ops(cql, mv1):
+    with pytest.raises(InvalidRequest, match='Cannot use'):
+        cql.execute(f'DROP TABLE {mv1}')
+    with pytest.raises(InvalidRequest, match='Cannot use'):
+        cql.execute(f"ALTER TABLE {mv1} WITH comment='hello'")
+    # Unlike the previous operations, in DESC TABLE cassandra doesn't explain
+    # that DESC TABLE cannot be used on a view and that DESC MATERIALIZED VIEW
+    # should be used instead - it just reports that the table is "not found".
+    # Reproduces #21026 (DESC TABLE was allowed on a view):
+    with pytest.raises(InvalidRequest):
+        cql.execute(f'DESC TABLE {mv1}')
+
+# A materialized view cannot have its own materialized views, nor secondary
+# indexes. Verify that.
+def test_view_forbids_view_or_index(cql, mv1):
+    # Cassandra and Scylla print different error messages here. Cassandra
+    # just claims that "Base table '{mv1}' doesn't exist" (it exists, but it's
+    # not a base table). Scylla says "Materialized views cannot be created
+    # against other materialized views". Let's allow both:
+    with pytest.raises(InvalidRequest, match='Base table|materialized views'):
```

Comment:
I don't feel like this match brings much value. I think it would be better to either remove it or change it to full error messages

Reply:
Let me explain the rationale.

I could do `pytest.raises(InvalidRequest)` and that's it. Sometimes I even do that. But then there's a risk: Imagine that in some future version of Scylla, we change something completely unrelated. For example, imagine that you are no longer allowed to write "p is not null" in the WHERE clause (it's already redundant, because p is a partition key so can't be null). So imagine the new_materialized_view() below suddenly throws a completely different error - say "InvalidRequest: 'p is not null' is not allowed" (I'm just making this up, but I hope you understand where I'm going). As soon as that happens, this test suddenly becomes a no-op - it will always pass even if we no longer produce an error on view-on-view.

This is why I always try to put even a fragment of the error message in the "match" - the silly "p is not null is not allowed" will not match even this fragment.

This is not a theoretical risk - I've seen this issue happening at least twice in recent years - and causing real bugs to be missed when the test continued to pass because it threw an exception - just not the right one.

I don't agree that having full error messages would be better. Some of these error messages are stupid and I hope one day they will be improved, and if they do, I wouldn't like many tests to suddenly fail. The purpose of this test isn't to "enshrine" the specific error messages that Scylla (or Cassandra) currently print - as I noted above the only purpose of this "match" is to avoid accepting completely different errors as valid.

Reply:
That's fair enough and I agree with it. My reasoning was that if the messages change, the commit changing the messages can also adjust this test; it runs every time in CI, so there's no harm in using a full error. On the contrary, it ensures that we don't catch an unrelated `InvalidRequest` that just happens to contain `"materialized views"` or `"Base table"` as a substring. I don't think it's critical, though (which is why I also suggested dropping the match altogether as an alternative solution), so I won't push it—I just wanted to elaborate on what I had in mind.

Reply:
You're absolutely right that it's less frustrating to have full error messages in in-tree tests (my frustration with full error messages in tests comes from the old days of dtest, that was out-of-tree).

But it's still a hassle to have to change tests just because you wanted to improve the error message. Imagine that you want to change some popular (but badly phrased) error message, and now you need to update 17 tests in areas you know nothing about - when none of these tests never had the goal of insisting that the error message never change. I agree that it's not critical either way given the tests are in-tree and that in practice, we don't often change error messages anyway.